### DISCUSSION TITLE
News on manifest V3 (MV3)

### DISCUSSION THREAD

**Discussion by WorldLanguages** [ScratchAddons/ScratchAddons#5162] [Discussion]
**EDIT: Latest blog post from the Chrome team:** https://developer.chrome.com/blog/resuming-the-transition-to-mv3/

Here's a summary:

EDIT: these dates are outdated. Chrome keeps changing them.

<s>**If we don't update to MV3 by ... the consequences will be:**
- **January 2023**: we might lose the "featured" badge, but we may continue to push MV2 updates.
- **June 2023**: the extension must be "unlisted", a portion of Chrome stable users will not be able to use MV2, but we may continue to push MV2 updates.
- **January 2024**: MV2 extensions are removed from the Chrome Web Store. We must push an MV3 update or disappear from the store.

Of course, we will update to MV3 before June 2023. But, we previously thought Chrome would stop supporting MV2 entirely by January 2023, and that's not the case. The "featured" tag is the only incentive until mid-2023.</s>

By the way, it's possible that no single `manifest.json` file will be accepted by both Chrome and Firefox at the same time. This would mean we would need to add a build step to the extension.
We currently do ship different `manifest.json` files to each browser store, but we use the same one for development. It looks like there will be no way to keep that situation on MV3, because Chrome expects `background.service_worker` to be defined (as long as the extension wants to have background code, and we do), and Firefox throws if it finds that property, because it decided to support "event pages" instead.
Supporting SWs and "event pages" at the same time isn't an issue, because SW code will work perfectly fine on an event page, but it would require two different `manifest.json` files for development, changing our current workflow.
I would need to research more deeply into this to have a final say, but it's likely a build step will be necessary. Testing a PR won't be as easy as downloading the zipball for that branch, and uploading it into the browser extensions page. Running a node.js command would be required.
Edit: see https://github.com/w3c/webextensions/issues/282

**Comment by Hans5958** [comment]
Reminder: Should a build step be required, GitHub Actions can build it for people to download from the website.

**Reply by WorldLanguages** [reply]
Thanks for the reminder. Of course, we could provide a way for users to download the built zip for a specific branch, without them needing nodejs (the action would run the nodejs command for them).
However, nodejs would still be required for day-to-day pull requests, in order to test them in the browser.

**Reply by Hans5958** [reply]
You could do the build workflow on every pull request. It's the one  when you need to build it for development.

**Reply by apple502j** [reply]
Doesn't sound that great, IMO. The "build step" mentioned here is just changing the manifest schema. Building vue is another question.

**Reply by Hans5958** [reply]
Eh, if build step is going to be essential for development, then I don't think it's worth to keep this old, no-building-required way for the sake of..., what?

**Comment by BroJac5246** [comment]
That's annoying, but there's really nothing we can do about it.

It would be good to target January 2023 for MV3, but that would disrupt our workflow earlier. However, the featured badge gives us extra trust with users, so ideally we hold onto it. What do you mean by "_might_ lose" the featured badge, though?

Anyways, we should probably just work on getting ready. The worst thing we could do is procrastinate.

**Reply by cobaltt7** [reply]
> However, the featured badge gives us extra trust with users, so ideally we hold onto it.

featured badge means nothing, there are featured malware extensions out there

**Reply by BroJac5246** [reply]
> > However, the featured badge gives us extra trust with users, so ideally we hold onto it.
> 
> featured badge means nothing, there are featured malware extensions out there

I know, but the word featured still gives some users a false sense of security (similar to the https lock). It might not actually mean much, but if some people think it does then it might be worth keeping.

**Comment by DNin01** [comment]
For people like me and some new contributors who don't use Git or npm yet, needing to build the extension could add extra complexity to contributing. It'd be nice if there could be a way to choose either to "manually" install the extension by keeping one of two `manifest.json` files or to build it. Each way has its own ups and downs, but there's probably a way to make either option work.

On the topic of when to update, IMO, it would be great if we could aim for an update to Manifest V3 by the end of January, just to get it out of the way. I'm not sure how hard that would be, though.

**Reply by WorldLanguages** [reply]
We could include pre-processed `manifest.json` files here on the repository, but the file would need to be renamed manually by the user.

When we decide to update depends on how https://github.com/w3c/webextensions/issues/282 advances. We must add a build step if we can't have the same `manifest.json` file accepted on the most modern versions of Chrome and Firefox at the same time.

**Reply by WorldLanguages** [reply]
It looks like there's some progress on supporting a unified manifest.json format. See link in above comment.

**Reply by DNin01** [reply]
I think by now we should just go ahead and split the manifest file in two, put them in a `src` or `components` directory or something, and add a build script and/or list of manual steps for assembling it for the browser that is being used. These steps would need to be followed no matter the browser as we have to `.gitignore` the output file.

Maybe we should make the entire project, including addons, be built, while we're at it? It could give us more options.

**Comment by DNin01** [comment]
Apparently, the Featured badge change is being postponed. [(Source)](https://developer.chrome.com/docs/extensions/mv3/mv2-sunset/)

**Reply by BroJac5246** [reply]
They also linked a simpler explanation [here](https://groups.google.com/u/1/a/chromium.org/g/chromium-extensions/c/zQ77HkGmK9E). Despite this postponement, we should still keep a focus on MV3 migration, however.

**Reply by WorldLanguages** [reply]
The way the Chrome Web Store has been handling MV3 is pretty ridiculous.

**Comment by BroJac5246** [comment]
> Of course, we will update to MV3 before June 2023.

Bump

**Reply by WorldLanguages** [reply]
All the deadlines have been postponed though.

**Reply by BroJac5246** [reply]
> All the deadlines have been postponed though.

I know but we're still behind where we should (ideally) be.

**Reply by WorldLanguages** [reply]
We're currently blocked by https://github.com/w3c/webextensions/issues/282 and it probably makes sense to wait until Chrome resolves that issue. The "Refined GitHub" extension is also waiting on the same issue.

**Comment by BroJac5246** [comment]
New update:
https://www.xda-developers.com/google-manifest-v3-extensions-update/

**Reply by WorldLanguages** [reply]
Thanks for the heads up!

**Reply by WorldLanguages** [reply]
Now that mv3 is actually mature, I expect this to be the real deadline.

When I find some time I'll create a time-line of tasks we need to do to upgrade. Most are pretty minor. The worst part for users will probably be dropping support for many old Chrome versions.

**Reply by DNin01** [reply]
> The worst part for users will probably be dropping support for many old Chrome versions.

On that note, we should probably notify those users somehow. Chrome probably won't push any updates to them past our future MV3 migration.

**Reply by WorldLanguages** [reply]
We will push MV2 updates that block these users from using the extension at least 1 month before releasing with MV3. But with the planned "unsupported browser" page, they'll be able to export their settings in the process.

I'll write an issue/discussion detailing these plans soon. There will also be updated requirements on which browsers will be supported for local development (we expect devs to have the latest stable versions though). Might also mean we could use some features such as CSS nesting from now on.

**Reply by DNin01** [reply]
From our conversation on #6751, this sounds fine to me.

**Comment by DNin01** [comment]
Apparently, Firefox is trying to [support service workers](https://bugzilla.mozilla.org/show_bug.cgi?id=1573659), if I'm right?

They also fixed a bug where background pages wouldn't work if `service_worker` was specified, in Firefox 121.

From https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/background#browser_support:

> Firefox:
> - background.service_worker is not supported (see [Firefox bug 1573659](https://bugzilla.mozilla.org/show_bug.cgi?id=1573659)).
> - supports background.scripts (or background.page) if service_worker is not specified or the service worker feature is disabled. Before Firefox 120, Firefox did not start the background page if service_worker was present (see [Firefox bug 1860304](https://bugzil.la/1860304)). From Firefox 121, the background page starts as expected, regardless of the presence of service_worker.

**Reply by WorldLanguages** [reply]
Yeah I asked for that bug to be fixed.

**Reply by WorldLanguages** [reply]
Chrome also did changes to their manifest JSON validator for better cross browser compatibility. Including a "scripts" property in the background object only logs a warning, instead of throwing an unavoidable error, in current Chrome Canary.
